File: frontend/src/types/policy.ts
export interface Policy {
  id: string;

  policy_number: string;

  policyholder_member_id: string;
  health_plan_id: string;

  status: string;
  policy_type: string;

  effective_date: string;
  expiry_date: string;

  network_name: string | null;

  annual_limit: string;
  deductible_amount: string;
  copay_amount: string;
  coinsurance_percentage: string;
  out_of_pocket_maximum: string;

  premium_amount: string;
  premium_currency: string;

  billing_frequency: string;
  billing_status: string;

  next_payment_date: string | null;

  benefits_summary: string | null;
  exclusions_summary: string | null;

  renewal_eligible: boolean;
  renewal_due_date: string | null;

  cancellation_reason: string | null;
  suspension_reason: string | null;

  is_active: boolean;

  created_at: string;
  updated_at: string;
}


export interface PolicyCreate {
  policy_number: string;

  policyholder_member_id: string;
  health_plan_id: string;

  status?: string;
  policy_type?: string;

  effective_date: string;
  expiry_date: string;

  network_name?: string | null;

  annual_limit?: number;
  deductible_amount?: number;
  copay_amount?: number;
  coinsurance_percentage?: number;
  out_of_pocket_maximum?: number;

  premium_amount?: number;
  premium_currency?: string;

  billing_frequency?: string;
  billing_status?: string;

  next_payment_date?: string | null;

  benefits_summary?: string | null;
  exclusions_summary?: string | null;

  renewal_eligible?: boolean;
  renewal_due_date?: string | null;

  cancellation_reason?: string | null;
  suspension_reason?: string | null;

  is_active?: boolean;
}


export interface PolicyUpdate {
  status?: string | null;
  policy_type?: string | null;

  effective_date?: string | null;
  expiry_date?: string | null;

  network_name?: string | null;

  annual_limit?: number | null;
  deductible_amount?: number | null;
  copay_amount?: number | null;
  coinsurance_percentage?: number | null;
  out_of_pocket_maximum?: number | null;

  premium_amount?: number | null;
  premium_currency?: string | null;

  billing_frequency?: string | null;
  billing_status?: string | null;

  next_payment_date?: string | null;

  benefits_summary?: string | null;
  exclusions_summary?: string | null;

  renewal_eligible?: boolean | null;
  renewal_due_date?: string | null;

  cancellation_reason?: string | null;
  suspension_reason?: string | null;

  is_active?: boolean | null;
}

File: backend/app/schemas/policy.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    model_validator,
)


class PolicyBase(BaseModel):
    policy_number: str = Field(
        min_length=3,
        max_length=100,
    )

    policyholder_member_id: UUID
    health_plan_id: UUID

    status: str = Field(
        default="PENDING",
        min_length=2,
        max_length=50,
    )

    policy_type: str = Field(
        default="INDIVIDUAL",
        min_length=2,
        max_length=50,
    )

    effective_date: date
    expiry_date: date

    network_name: str | None = Field(
        default=None,
        max_length=255,
    )

    annual_limit: Decimal = Field(
        default=0,
        ge=0,
    )

    deductible_amount: Decimal = Field(
        default=0,
        ge=0,
    )

    copay_amount: Decimal = Field(
        default=0,
        ge=0,
    )

    coinsurance_percentage: Decimal = Field(
        default=0,
        ge=0,
        le=100,
    )

    out_of_pocket_maximum: Decimal = Field(
        default=0,
        ge=0,
    )

    premium_amount: Decimal = Field(
        default=0,
        ge=0,
    )

    premium_currency: str = Field(
        default="SAR",
        min_length=3,
        max_length=10,
    )

    billing_frequency: str = Field(
        default="MONTHLY",
        min_length=2,
        max_length=50,
    )

    billing_status: str = Field(
        default="PENDING",
        min_length=2,
        max_length=50,
    )

    next_payment_date: date | None = None

    benefits_summary: str | None = None
    exclusions_summary: str | None = None

    renewal_eligible: bool = False
    renewal_due_date: date | None = None

    cancellation_reason: str | None = None
    suspension_reason: str | None = None

    is_active: bool = True

    @model_validator(mode="after")
    def validate_policy_dates(
        self,
    ) -> "PolicyBase":
        if self.expiry_date < self.effective_date:
            raise ValueError(
                "Policy expiry date cannot be earlier "
                "than the effective date.",
            )

        if (
            self.renewal_due_date is not None
            and self.renewal_due_date
            < self.effective_date
        ):
            raise ValueError(
                "Renewal due date cannot be earlier "
                "than the policy effective date.",
            )

        return self


class PolicyCreate(PolicyBase):
    pass


class PolicyUpdate(BaseModel):
    status: str | None = Field(
        default=None,
        min_length=2,
        max_length=50,
    )

    policy_type: str | None = Field(
        default=None,
        min_length=2,
        max_length=50,
    )

    effective_date: date | None = None
    expiry_date: date | None = None

    network_name: str | None = Field(
        default=None,
        max_length=255,
    )

    annual_limit: Decimal | None = Field(
        default=None,
        ge=0,
    )

    deductible_amount: Decimal | None = Field(
        default=None,
        ge=0,
    )

    copay_amount: Decimal | None = Field(
        default=None,
        ge=0,
    )

    coinsurance_percentage: Decimal | None = Field(
        default=None,
        ge=0,
        le=100,
    )

    out_of_pocket_maximum: Decimal | None = Field(
        default=None,
        ge=0,
    )

    premium_amount: Decimal | None = Field(
        default=None,
        ge=0,
    )

    premium_currency: str | None = Field(
        default=None,
        min_length=3,
        max_length=10,
    )

    billing_frequency: str | None = Field(
        default=None,
        min_length=2,
        max_length=50,
    )

    billing_status: str | None = Field(
        default=None,
        min_length=2,
        max_length=50,
    )

    next_payment_date: date | None = None

    benefits_summary: str | None = None
    exclusions_summary: str | None = None

    renewal_eligible: bool | None = None
    renewal_due_date: date | None = None

    cancellation_reason: str | None = None
    suspension_reason: str | None = None

    is_active: bool | None = None


class PolicyResponse(PolicyBase):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: UUID
    created_at: datetime
    updated_at: datetime